docs/rfcs/017-timeline-data-management.md

```diff
@@ -0,0 +1,409 @@
+- `freeze_inmem_layer(true)` that touches both `write_lock` and `layer_flush_lock` seems not very aligned with the rest of the locks to those primitives; it also now restricts the layer creation concurrency even more, yet there are various `freeze_inmem_layer(false)` that are ignoring those restrictions at the same time
+
+![timeline data access synchronization(2)](./images/017-timeline-data-management/timeline_data_access_sync_2.svg)
+
+Comments:
+
+- `partitioning: Mutex<(KeyPartitioning, Lsn)>` lock is a data sync lock that's not used to synchronize the tasks (all other such kinds were considered "almost instant" and omitted on the diagram), yet is very similar to what `write_lock` and `layer_flush_lock` do: it ensures the timeline in-memory data is up-to-date with the layer files state on disk, which is what `LayerMap` is for.
+
+- there are multiple locks that do similar task management operations:
+  - `gc_cs: Mutex<()>` and `latest_gc_cutoff_lsn: RwLock<Lsn>` ensures that branching and gc are not run concurrently
+  - `layer_removal_cs: Mutex<()>` lock ensure gc, compaction and timeline deletion via HTTP API do not run concurrently
+  - `file_lock: RwLock<()>` is used as a semaphore, to ensure "all" gc and compaction tasks are shut down and do not start
+    Yet that lock does take only gc and compaction from internal loops: libpq call is not cancelled and waited upon.
+
+Those operations do not seem to belong to a timeline. Moreover, some of those could be eliminated entirely due to duplication of their tasks.
+
+## Proposed implementation
+
+### How to structure timeline data access better
+
+- adjust tenant state handling
+
+Current [`TenantState`](https://github.com/neondatabase/neon/blob/28243d68e60ffc7e69f158522f589f7d2e09186d/pageserver/src/tenant_mgr.rs#L108) [changes](https://github.com/neondatabase/neon/blob/28243d68e60ffc7e69f158522f589f7d2e09186d/pageserver/src/tenant_mgr.rs#L317) mainly indicates whether GC and compaction tasks are running or not; another state, `Broken` shows only in case any timeline does not load during startup.
+
+We could start both GC and compaction tasks at the time the tenant is created and adjust the tasks to throttle/sleep on timeline absence and wake up when the first one is added.
+The latter becomes more important on download on demand, since we won't have the entire timeline in reach to verify its correctness. Moreover, if any network connection happens, the timeline could fail temporarily and entire tenant should be marked as broken due to that.
+
+Since nothing verifies the `TenantState` via HTTP API currently, it makes sense to remove the whole state entirely and don't write the code to synchronize its changes.
+Instead, we could indicate internal issues for every timeline and have a better API to "stop" timeline processing without deleting its data, making our API less restrictive.
```

Comment:
I think we'll still need something like the TenantState enum. When an 'attach_tenant' request comes in, I presume you'll first create an in-memory object to represent the tenant, and then start loading its state from S3. Even if the layer files are downloaded on-demand, you still need to download information about the timelines before you can start accepting GetPage requests or do much else with the tenant.

If a GetPage request comes in for a tenant, while it's still been loaded, the request should either wait for the tenant-attach operation to finish, or return an error. Also if another Attach request for the same tenant comes in, you need to somehow notice that the tenant is already been loaded, and wait for it to finish or throw an error.

So we will need to store information about tenants that are being loaded, somewhere in-memory. It seems most strraightforward to have a 'state' field in the Tenant or Repository struct, like the current TenantState. 

Similarly, when a tenant is being detached, you need a shutdown sequence where you prevent new processing new WAL for that tenant, but continue uploading everything to S3. You'll need some kind of a "shutting down" state for that.

Reply:
> I think we'll still need something like the TenantState enum

Have mentioned that while adding the `Tenant` object struct, indeed some sync could be used there.
I'm not sure how to structure that bit, since similar questions [arise](https://github.com/neondatabase/neon/pull/2152#discussion_r930807532).

There has to be something, that's able to throttle tasks and global events, while communicating with the logic that processes HTTP.

Could be the place we store `Tenant` data in.